File: apps/worker/src/processor.ts
import { db } from "@silo/database";
import * as schema from "@silo/database/schema";
import { config } from "@silo/engine/config";
import {
  commitOffsetsRest,
  produceRecordRest,
  type RestConsumerInstance,
  type RestRecord,
} from "@silo/engine/kafka/rest-client";
import { getHandlerForTopic } from "./handlers/topic-handlers";

export const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

function getMessageId(payload: unknown): string | undefined {
  const payloadObj = (payload as Record<string, unknown> | null) ?? null;
  const source = payloadObj?.source as Record<string, unknown> | undefined;
  const messageId =
    payloadObj?.message_id ??
    payloadObj?.messageId ??
    payloadObj?.id ??
    source?.messageId;
  return typeof messageId === "undefined" ? undefined : String(messageId);
}

async function commitNextOffset(
  instance: RestConsumerInstance,
  record: RestRecord,
): Promise<void> {
  await commitOffsetsRest(instance, [
    {
      topic: record.topic,
      partition: record.partition,
      offset: (Number(record.offset) + 1).toString(),
    },
  ]);
}

async function sendToDlq(topic: string, raw: string, key?: string): Promise<void> {
  await produceRecordRest(`${config.kafka.dlqPrefix}${topic}`, raw, key);
}

async function handleInvalidRecord(
  instance: RestConsumerInstance,
  record: RestRecord,
  raw: string,
  reason: string,
): Promise<void> {
  console.error(
    `[KAFKA-REST] ${reason}, sending to DLQ`,
    record.topic,
  );
  try {
    await sendToDlq(record.topic, raw);
  } catch (error) {
    console.error("[KAFKA-REST] failed to send to DLQ", error);
  }
  try {
    await commitNextOffset(instance, record);
  } catch (error) {
    console.error("[KAFKA-REST] commitOffsets failed", error);
  }
}

export async function processRecord(
  instance: RestConsumerInstance,
  record: RestRecord,
): Promise<void> {
  const raw =
    typeof record.value === "string"
      ? record.value
      : JSON.stringify(record.value);

  let payload: unknown = null;
  try {
    payload = raw.length > 0 ? JSON.parse(raw) : null;
  } catch (error) {
    console.error("[KAFKA-REST] invalid JSON payload", error);
    await handleInvalidRecord(
      instance,
      record,
      raw,
      "invalid JSON payload",
    );
    return;
  }

  const messageId = getMessageId(payload);
  if (!messageId) {
    await handleInvalidRecord(
      instance,
      record,
      raw,
      "message without message_id/source.messageId",
    );
    return;
  }

  const handler = getHandlerForTopic(record.topic);
  const maxAttempts = config.kafka.processRetryCount || 3;
  const baseBackoff = config.kafka.retryBackoffMs || 1000;
  let attempt = 0;
  let succeeded = false;

  while (attempt < maxAttempts && !succeeded) {
    attempt += 1;
    try {
      await db.transaction(async (tx) => {
        try {
          await tx.insert(schema.kafkaProcessedMessages).values({
            topic: record.topic,
            messageId,
            handler: handler.name || record.topic,
          });
        } catch (error: unknown) {
          if (
            typeof error === "object" &&
            error !== null &&
            "code" in error &&
            (error as { code?: unknown }).code === "23505"
          ) {
            succeeded = true;
            return;
          }
          throw error;
        }

        await handler({
          topic: record.topic,
          partition: record.partition,
          messageId,
          payload,
          tx,
        });
      });

      succeeded = true;
    } catch (error) {
      console.error(
        `[KAFKA-REST] error processing message (attempt ${attempt})`,
        error,
      );
      if (attempt < maxAttempts) {
        await sleep(baseBackoff * Math.pow(2, attempt - 1));
      }
    }
  }

  if (!succeeded) {
    try {
      await sendToDlq(record.topic, raw, messageId);
    } catch (error) {
      console.error("[KAFKA-REST] failed to send to DLQ", error);
    }
  }

  try {
    await commitNextOffset(instance, record);
  } catch (error) {
    console.error("[KAFKA-REST] failed to commit offset", error);
  }
}

File: apps/backend/src/silo/worker/processor.py
from __future__ import annotations

import asyncio
import json
import logging
from collections.abc import Mapping
from typing import Any

from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.dialects.sqlite import insert as sqlite_insert
from sqlalchemy.engine import Connection, Engine
from sqlalchemy.exc import IntegrityError

from silo.db.models import legacy_tables
from silo.services.kafka_rest import KafkaRestClient, RestConsumerInstance
from silo.worker.config import WorkerSettings
from silo.worker.handlers.topic_handlers import get_handler_for_topic
from silo.worker.health import WorkerHealthMonitor

logger = logging.getLogger(__name__)
sleep = asyncio.sleep


def normalize_record_value(value: object) -> str:
    if isinstance(value, str):
        return value
    if value is None:
        return "null"
    if isinstance(value, bytes):
        return value.decode("utf-8", errors="replace")
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), default=str)


def parse_record_value(value: object) -> object:
    if isinstance(value, str):
        return json.loads(value)
    return value


def extract_message_id(payload: object) -> str | None:
    if not isinstance(payload, Mapping):
        return None

    for key in ("message_id", "messageId", "id"):
        value = payload.get(key)
        text = _optional_text(value)
        if text is not None:
            return text

    source = payload.get("source")
    if isinstance(source, Mapping):
        text = _optional_text(source.get("messageId"))
        if text is not None:
            return text

    return None


async def process_record(
    *,
    engine: Engine,
    client: KafkaRestClient,
    instance: RestConsumerInstance,
    record: Mapping[str, Any],
    settings: WorkerSettings,
    health: WorkerHealthMonitor | None = None,
) -> None:
    topic = str(record.get("topic") or "")
    partition = int(record.get("partition") or 0)
    offset = str(record.get("offset") or "0")
    raw = normalize_record_value(record.get("value"))

    try:
        payload = parse_record_value(record.get("value"))
    except Exception:
        logger.exception("[KAFKA-REST] invalid JSON payload, sending to DLQ")
        await _handle_invalid_record(
            client,
            instance,
            topic=topic,
            partition=partition,
            raw=raw,
            offset=offset,
            settings=settings,
        )
        return

    message_id = extract_message_id(payload)
    if not message_id:
        logger.error("[KAFKA-REST] message without message_id/source.messageId, sending to DLQ")
        await _handle_invalid_record(
            client,
            instance,
            topic=topic,
            partition=partition,
            raw=raw,
            offset=offset,
            settings=settings,
        )
        return

    handler = get_handler_for_topic(topic)
    handler_name = (
        getattr(handler, "__worker_handler_name__", None)
        or getattr(handler, "__name__", "")
        or topic
    )
    max_attempts = settings.process_retry_count or 3
    base_backoff_ms = settings.retry_backoff_ms or 1_000
    attempt = 0
    succeeded = False

    while attempt < max_attempts and not succeeded:
        attempt += 1
        try:
            with engine.begin() as connection:
                inserted = _insert_processed_message(
                    connection,
                    topic=topic,
                    message_id=message_id,
                    handler_name=handler_name,
                )
                if not inserted:
                    succeeded = True
                    break

                await handler(
                    topic=topic,
                    partition=partition,
                    message_id=message_id,
                    payload=payload,
                    connection=connection,
                )
            succeeded = True
        except Exception:
            logger.exception("[KAFKA-REST] error processing message (attempt %s)", attempt)
            if attempt < max_attempts:
                await sleep(base_backoff_ms * (2 ** (attempt - 1)) / 1000.0)

    if not succeeded:
        try:
            await client.produce_record_rest(f"{settings.dlq_prefix}{topic}", raw, message_id)
        except Exception:
            logger.exception("[KAFKA-REST] failed to send to DLQ")
            if health is not None:
                health.mark_error("dlq_failed")
            return

    try:
        await client.commit_offsets_rest(
            instance,
            [
                {
                    "topic": topic,
                    "partition": partition,
                    "offset": str(int(offset) + 1),
                }
            ],
        )
    except Exception:
        logger.exception("[KAFKA-REST] failed to commit offset")
        if health is not None:
            health.mark_error("commit_failed")


def _insert_processed_message(
    connection: Connection,
    *,
    topic: str,
    message_id: str,
    handler_name: str,
) -> bool:
    table = legacy_tables["kafka_processed_messages"]
    values = {
        "topic": topic,
        "message_id": message_id,
        "handler": handler_name,
    }
    if connection.dialect.name == "postgresql":
        statement = (
            pg_insert(table)
            .values(values)
            .on_conflict_do_nothing(index_elements=["topic", "message_id"])
            .returning(table.c.topic)
        )
    elif connection.dialect.name == "sqlite":
        statement = (
            sqlite_insert(table)
            .values(values)
            .on_conflict_do_nothing(index_elements=["topic", "message_id"])
            .returning(table.c.topic)
        )
    else:  # pragma: no cover - fallback for unsupported dialects
        statement = table.insert().values(values)

    try:
        result = connection.execute(statement)
    except IntegrityError:
        return False

    if connection.dialect.name in {"postgresql", "sqlite"}:
        return result.first() is not None
    return True


async def _handle_invalid_record(
    client: KafkaRestClient,
    instance: RestConsumerInstance,
    *,
    topic: str,
    partition: int,
    raw: str,
    offset: str,
    settings: WorkerSettings,
) -> None:
    try:
        await client.produce_record_rest(f"{settings.dlq_prefix}{topic}", raw, None)
    except Exception:
        logger.exception("[KAFKA-REST] failed to send invalid record to DLQ")
        return

    try:
        await client.commit_offsets_rest(
            instance,
            [{"topic": topic, "partition": partition, "offset": str(int(offset) + 1)}],
        )
    except Exception:
        logger.exception("[KAFKA-REST] failed to commit offset after invalid record")


def _optional_text(value: object | None) -> str | None:
    if isinstance(value, str):
        text = value.strip()
        return text or None
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return str(value)
    return None
